File: apps/agent/src/agent/extract-graph.ts
import { ObjectId } from "mongodb";
import { generate } from "../lib/vertex.js";
import { getCollections } from "../lib/mongo.js";
import {
  clearGraph,
  entityKey,
  insertRelation,
  upsertEntity,
  type EntityKind,
  type RelationRecord,
} from "../lib/graph.js";

export type GraphEvent =
  | { kind: "start"; totalChunks: number; at: number }
  | { kind: "batch"; index: number; total: number; chunksInBatch: number; entitiesFound: number; relationsFound: number; at: number }
  | { kind: "done"; totalEntities: number; totalRelations: number; totalMs: number; at: number }
  | { kind: "error"; message: string; at: number };

const SYSTEM = `You extract a structured memory graph from a senior PM's corpus (emails, calendar, meeting notes, shared docs, slack, personal jots).

For each batch of chunks, return STRICT JSON — no markdown, no preamble — with this exact shape:

{
  "entities": [
    {
      "name": "Sarah Okafor",
      "kind": "person" | "project" | "topic",
      "role": "Director of Eng, Lantern team",
      "chunkIds": ["..."]
    }
  ],
  "relations": [
    {
      "from": "Alex Chen",
      "to": "Sarah Okafor",
      "kind": "owes" | "works_with" | "manages" | "discusses",
      "evidence": "verbatim phrase from the chunk",
      "chunkId": "..."
    }
  ]
}

Rules:
- Only extract entities that appear with a clear name (proper noun). Skip pronouns, anaphoric refs, vague descriptors.
- "person" = a named human. "project" = a named product / initiative / workstream. "topic" = a named theme that recurs across multiple chunks (rare — be conservative).
- Use canonical full names (e.g. "Sarah Okafor" not "Sarah", unless the chunk only says "Sarah").
- "role" should be ONE short clause derived from the chunks — what the entity is or does. Skip if unclear.
- "owes" relations require a concrete commitment: "X will deliver Y by Z", "X promised Y", "AC: X to follow up on Y".
- "works_with" is for collaboration ties evidenced by joint meetings / cc / co-authored docs.
- "manages" is reporting lines.
- "discusses" is the weakest tie — use only when stronger types don't fit.
- chunkIds in entities = list of chunkIds from THIS batch where the entity appears.
- chunkId in relations = single chunkId where the evidence lives.
- Do NOT hallucinate. If a chunk doesn't have a clear entity or relation, leave it out.
- Output JSON only. First character must be '{'.`;

interface ExtractInput {
  chunkId: string;
  source: string;
  title: string;
  date?: string;
  text: string;
}

interface ExtractResponse {
  entities?: Array<{
    name?: unknown;
    kind?: unknown;
    role?: unknown;
    chunkIds?: unknown;
  }>;
  relations?: Array<{
    from?: unknown;
    to?: unknown;
    kind?: unknown;
    evidence?: unknown;
    chunkId?: unknown;
  }>;
}

const BATCH_SIZE = 12;
const RELEVANT_SOURCES = new Set(["email", "calendar", "meeting_notes", "shared_doc", "slack", "notes"]);

export async function* runGraphExtraction({ rebuild }: { rebuild: boolean }): AsyncGenerator<GraphEvent, void, unknown> {
  const startedAt = Date.now();

  try {
    if (rebuild) await clearGraph();

    const { chunks } = await getCollections();
    const filter = { source: { $in: Array.from(RELEVANT_SOURCES) as unknown as ["email"] } };
    const all = await chunks
      .find(filter, {
        projection: {
          _id: 1,
          source: 1,
          title: 1,
          text: 1,
          "metadata.date": 1,
          "metadata.eventTime": 1,
        },
        sort: { "metadata.date": 1, "metadata.eventTime": 1 },
      })
      .toArray();

    const inputs: ExtractInput[] = all.map((c) => ({
      chunkId: String(c._id),
      source: String(c.source),
      title: String(c.title),
      text: typeof c["text"] === "string" ? (c["text"] as string) : "",
      date:
        typeof c["metadata"] === "object" && c["metadata"]
          ? (((c["metadata"] as Record<string, unknown>)["date"] as string | undefined) ??
            ((c["metadata"] as Record<string, unknown>)["eventTime"] as string | undefined))
          : undefined,
    }));

    yield { kind: "start", totalChunks: inputs.length, at: Date.now() };

    // Accumulator: collapse duplicate entities across batches and update aggregate fields.
    const entityAgg = new Map<
      string,
      {
        name: string;
        kind: EntityKind;
        role: string | undefined;
        chunkIds: Set<string>;
        dates: string[];
      }
    >();

    let totalRelations = 0;
    const batches: ExtractInput[][] = [];
    for (let i = 0; i < inputs.length; i += BATCH_SIZE) {
      batches.push(inputs.slice(i, i + BATCH_SIZE));
    }

    for (let b = 0; b < batches.length; b++) {
      const batch = batches[b]!;
      const prompt = buildBatchPrompt(batch);

      const r = await generate({
        system: SYSTEM,
        prompt,
        temperature: 0.1,
        maxTokens: 8192,  // Gemini 3 consumes thinking tokens regardless of budget — leave headroom for entity-dense batches
        responseMimeType: "application/json",
        thinkingBudget: 0,
      });

      let parsed: ExtractResponse = {};
      try {
        const text = r.text.trim();
        const start = text.indexOf("{");
        const end = text.lastIndexOf("}");
        if (start >= 0 && end > start) {
          parsed = JSON.parse(text.slice(start, end + 1)) as ExtractResponse;
        }
      } catch {
        // batch skipped on parse failure
      }

      const ents = Array.isArray(parsed.entities) ? parsed.entities : [];
      const rels = Array.isArray(parsed.relations) ? parsed.relations : [];

      let entsAdded = 0;
      for (const e of ents) {
        const name = typeof e.name === "string" ? e.name.trim() : "";
        if (!name) continue;
        const kind =
          e.kind === "person" || e.kind === "project" || e.kind === "topic"
            ? e.kind
            : null;
        if (!kind) continue;
        const role = typeof e.role === "string" ? e.role.trim() : undefined;
        const chunkIds = Array.isArray(e.chunkIds)
          ? e.chunkIds.filter((c): c is string => typeof c === "string")
          : [];
        const key = entityKey(name);
        if (!key) continue;
        entsAdded++;
        const dates: string[] = chunkIds
          .map((cid) => batch.find((bi) => bi.chunkId === cid)?.date)
          .filter((d): d is string => typeof d === "string");
        const existing = entityAgg.get(key + "::" + kind);
        if (existing) {
          for (const cid of chunkIds) existing.chunkIds.add(cid);
          existing.dates.push(...dates);
          // prefer a longer / more specific role
          if (role && (!existing.role || role.length > existing.role.length)) {
            existing.role = role;
          }
        } else {
          entityAgg.set(key + "::" + kind, {
            name,
            kind,
            role,
            chunkIds: new Set(chunkIds),
            dates,
          });
        }
      }

      let relsAdded = 0;
      for (const rel of rels) {
        const from = typeof rel.from === "string" ? rel.from.trim() : "";
        const to = typeof rel.to === "string" ? rel.to.trim() : "";
        const kind = (rel.kind === "owes" || rel.kind === "works_with" || rel.kind === "manages" || rel.kind === "discusses")
          ? rel.kind as RelationRecord["kind"]
          : null;
        const evidence = typeof rel.evidence === "string" ? rel.evidence.trim() : "";
        const chunkIdStr = typeof rel.chunkId === "string" ? rel.chunkId : "";
        if (!from || !to || !kind || !evidence) continue;
        const fromKey = entityKey(from);
        const toKey = entityKey(to);
        if (!fromKey || !toKey || fromKey === toKey) continue;
        const date = chunkIdStr ? batch.find((bi) => bi.chunkId === chunkIdStr)?.date : undefined;
        try {
          await insertRelation({
            from: fromKey,
            to: toKey,
            kind,
            evidence,
            ...(chunkIdStr ? { chunkId: chunkIdStr } : {}),
            ...(date ? { date } : {}),
          });
          relsAdded++;
        } catch {
          // best-effort
        }
      }
      totalRelations += relsAdded;

      yield {
        kind: "batch",
        index: b + 1,
        total: batches.length,
        chunksInBatch: batch.length,
        entitiesFound: entsAdded,
        relationsFound: relsAdded,
        at: Date.now(),
      };
    }

    // Flush accumulated entities to Mongo. Build sparkline series along the way.
    for (const e of entityAgg.values()) {
      const sorted = e.dates
        .map((d) => isoDay(d))
        .filter((d): d is string => !!d)
        .sort();
      const seriesMap = new Map<string, number>();
      for (const d of sorted) seriesMap.set(d, (seriesMap.get(d) ?? 0) + 1);
      const series = Array.from(seriesMap.entries())
        .map(([date, count]) => ({ date, count }))
        .sort((a, b) => a.date.localeCompare(b.date));
      const firstSeen = sorted[0];
      const lastSeen = sorted[sorted.length - 1];

      try {
        await upsertEntity({
          name: e.name,
          key: entityKey(e.name),
          kind: e.kind,
          role: e.role,
          mentions: e.chunkIds.size,
          ...(firstSeen ? { firstSeen } : {}),
          ...(lastSeen ? { lastSeen } : {}),
          chunkIds: Array.from(e.chunkIds),
          series,
        });
      } catch {
        // best-effort
      }
    }

    yield {
      kind: "done",
      totalEntities: entityAgg.size,
      totalRelations,
      totalMs: Date.now() - startedAt,
      at: Date.now(),
    };
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    yield { kind: "error", message: msg, at: Date.now() };
  }
}

function buildBatchPrompt(batch: ExtractInput[]): string {
  const body = batch
    .map((c) => `--- chunkId: ${c.chunkId} | ${c.source} | ${c.title} ${c.date ? "| " + c.date : ""} ---\n${c.text}`)
    .join("\n\n");
  return `Extract entities and relations from these ${batch.length} chunks. Return JSON only.\n\n${body}`;
}

function isoDay(raw: string | undefined): string | null {
  if (!raw) return null;
  const d = new Date(raw);
  if (Number.isNaN(d.getTime())) return null;
  return d.toISOString().slice(0, 10);
}

// avoid unused import warning when ObjectId isn't referenced elsewhere
void ObjectId;

File: apps/agent-py/app/agent/extract_graph.py
"""Graph extraction — port of apps/agent/src/agent/extract-graph.ts."""
from __future__ import annotations

import json
import time
from collections import defaultdict

from app.db.mongo import chunks
from app.lib.graph import clear_graph, entity_key, insert_relation, upsert_entity
from app.llm.genai_client import generate

SYSTEM = """You extract a structured memory graph from a senior PM's corpus (emails, calendar, meeting notes, shared docs, slack, personal jots).

For each batch of chunks, return STRICT JSON — no markdown, no preamble — with this exact shape:

{
  "entities": [
    {"name": "Sarah Okafor", "kind": "person" | "project" | "topic", "role": "Director of Eng, Lantern team", "chunkIds": ["..."]}
  ],
  "relations": [
    {"from": "Alex Chen", "to": "Sarah Okafor", "kind": "owes" | "works_with" | "manages" | "discusses", "evidence": "verbatim phrase", "chunkId": "..."}
  ]
}

Rules:
- Only extract entities that appear with a clear name (proper noun). Skip pronouns/vague descriptors.
- "person" = a named human. "project" = a named product/initiative. "topic" = a named recurring theme (rare).
- Use canonical full names. "role" = ONE short clause derived from the chunks; skip if unclear.
- "owes" requires a concrete commitment. "works_with" = collaboration. "manages" = reporting lines. "discusses" = weakest tie.
- chunkIds in entities = chunkIds from THIS batch where the entity appears. chunkId in relations = single chunkId.
- Do NOT hallucinate. Output JSON only; first character '{'."""

BATCH_SIZE = 12
RELEVANT = ["email", "calendar", "meeting_notes", "shared_doc", "slack", "notes"]
_REL_KINDS = ("owes", "works_with", "manages", "discusses")


def _now() -> int:
    return int(time.time() * 1000)


def _iso_day(raw):
    if not raw:
        return None
    try:
        from datetime import datetime
        return datetime.fromisoformat(str(raw).replace("Z", "+00:00")).date().isoformat()
    except ValueError:
        return None


async def run_graph_extraction(*, rebuild: bool):
    started = time.time()
    try:
        if rebuild:
            await clear_graph()
        rows = await chunks().find(
            {"source": {"$in": RELEVANT}},
            projection={"_id": 1, "source": 1, "title": 1, "text": 1, "metadata.date": 1, "metadata.eventTime": 1},
            sort=[("metadata.date", 1), ("metadata.eventTime", 1)],
        ).to_list(length=None)
        inputs = [{
            "chunkId": str(c["_id"]), "source": str(c["source"]), "title": str(c.get("title", "")),
            "text": c.get("text") if isinstance(c.get("text"), str) else "",
            "date": (c.get("metadata") or {}).get("date") or (c.get("metadata") or {}).get("eventTime"),
        } for c in rows]

        yield {"kind": "start", "totalChunks": len(inputs), "at": _now()}

        # entity_agg[key::kind] = {name, kind, role, chunkIds:set, dates:list}
        entity_agg: dict[str, dict] = {}
        total_relations = 0
        batches = [inputs[i:i + BATCH_SIZE] for i in range(0, len(inputs), BATCH_SIZE)]

        for b, batch in enumerate(batches):
            body = "\n\n".join(
                f"--- chunkId: {c['chunkId']} | {c['source']} | {c['title']} "
                f"{('| ' + c['date']) if c['date'] else ''} ---\n{c['text']}" for c in batch)
            prompt = f"Extract entities and relations from these {len(batch)} chunks. Return JSON only.\n\n{body}"
            r = await generate(prompt, system=SYSTEM, temperature=0.1, max_tokens=8192,
                               response_mime_type="application/json", thinking_budget=0)
            parsed = {}
            try:
                t = r.text.strip()
                s, e = t.find("{"), t.rfind("}")
                if s >= 0 and e > s:
                    parsed = json.loads(t[s:e + 1])
            except json.JSONDecodeError:
                parsed = {}

            ents_added = 0
            for ent in (parsed.get("entities") or []):
                name = (ent.get("name") or "").strip() if isinstance(ent.get("name"), str) else ""
                kind = ent.get("kind") if ent.get("kind") in ("person", "project", "topic") else None
                if not name or not kind:
                    continue
                role = ent.get("role").strip() if isinstance(ent.get("role"), str) else None
                chunk_ids = [c for c in (ent.get("chunkIds") or []) if isinstance(c, str)]
                key = entity_key(name)
                if not key:
                    continue
                ents_added += 1
                dates = [bi["date"] for cid in chunk_ids for bi in batch if bi["chunkId"] == cid and bi["date"]]
                k = f"{key}::{kind}"
                if k in entity_agg:
                    ex = entity_agg[k]
                    ex["chunkIds"].update(chunk_ids)
                    ex["dates"].extend(dates)
                    if role and (not ex["role"] or len(role) > len(ex["role"])):
                        ex["role"] = role
                else:
                    entity_agg[k] = {"name": name, "kind": kind, "role": role,
                                     "chunkIds": set(chunk_ids), "dates": list(dates)}

            rels_added = 0
            for rel in (parsed.get("relations") or []):
                frm = (rel.get("from") or "").strip() if isinstance(rel.get("from"), str) else ""
                to = (rel.get("to") or "").strip() if isinstance(rel.get("to"), str) else ""
                kind = rel.get("kind") if rel.get("kind") in _REL_KINDS else None
                evidence = (rel.get("evidence") or "").strip() if isinstance(rel.get("evidence"), str) else ""
                chunk_id = rel.get("chunkId") if isinstance(rel.get("chunkId"), str) else ""
                if not (frm and to and kind and evidence):
                    continue
                fk, tk = entity_key(frm), entity_key(to)
                if not fk or not tk or fk == tk:
                    continue
                date = next((bi["date"] for bi in batch if bi["chunkId"] == chunk_id), None) if chunk_id else None
                rec = {"from": fk, "to": tk, "kind": kind, "evidence": evidence}
                if chunk_id:
                    rec["chunkId"] = chunk_id
                if date:
                    rec["date"] = date
                try:
                    await insert_relation(rec)
                    rels_added += 1
                except Exception:  # noqa: BLE001
                    pass
            total_relations += rels_added
            yield {"kind": "batch", "index": b + 1, "total": len(batches), "chunksInBatch": len(batch),
                   "entitiesFound": ents_added, "relationsFound": rels_added, "at": _now()}

        for ent in entity_agg.values():
            sorted_days = sorted(d for d in (_iso_day(d) for d in ent["dates"]) if d)
            series_map: dict[str, int] = defaultdict(int)
            for d in sorted_days:
                series_map[d] += 1
            series = [{"date": d, "count": n} for d, n in sorted(series_map.items())]
            try:
                await upsert_entity({
                    "name": ent["name"], "key": entity_key(ent["name"]), "kind": ent["kind"],
                    "role": ent["role"], "mentions": len(ent["chunkIds"]),
                    "firstSeen": sorted_days[0] if sorted_days else None,
                    "lastSeen": sorted_days[-1] if sorted_days else None,
                    "chunkIds": list(ent["chunkIds"]), "series": series,
                })
            except Exception:  # noqa: BLE001
                pass

        yield {"kind": "done", "totalEntities": len(entity_agg), "totalRelations": total_relations,
               "totalMs": int((time.time() - started) * 1000), "at": _now()}
    except Exception as err:  # noqa: BLE001
        yield {"kind": "error", "message": str(err), "at": _now()}
